Add cppdemangle in druntime

ErnyTech opened the pull request:
Add cppdemangle in druntime

Having a C++ symbol demangler is very useful, with this it will
be possible to implement the C++ symbol demangle in the stacktrace.

The demangling is done in Posix platform by function __cxa_demangle
presents in Itanium C++ ABI (https://itanium-cxx-abi.github.io/cxx-abi/abi.html#demangler).

Itanium C++ ABI is used practically in all modern C++ compilers on Posix,
however old compilers (like GCC < 3.0) used a different name mangling,
this is not a problem because DMD in fact only supports Itanium C++ ABI on Posix.

For Windows instead the implementation it is provided by the UnDecorateSymbolName
function present in the Debug Help Library.
(https://docs.microsoft.com/en-us/windows/win32/api/dbghelp/nf-dbghelp-undecoratesymbolname)

ErnyTech commented:
It is now ready for review.

Geod24 commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,268 @@
+        version (Posix)
+        {
+            int result;
+            auto demangledNamePtr = _instance.__cxa_demangle(mangledName, null, null, &result);
```
I don't see the point of copying the result when you could just pass the pointer to the array here ?

Geod24 reviewed (changes requested):
Did a first round of review

ErnyTech commented:
Thanks for the extensive review!

ErnyTech commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,268 @@
+
+version (Windows)
+{
+    version (ANSI) {} else version = Unicode;
```
See this https://github.com/dlang/druntime/blob/2eec30b35bab308a37298331353bdce5fee1b657/src/core/sys/windows/dbghelp_types.d#L16

https://github.com/dlang/druntime/blob/2eec30b35bab308a37298331353bdce5fee1b657/src/core/sys/windows/winnt.d#L56

I need it to know when to convert utf-8 (char) to utf-16 (wchar) for windows api

Geod24 replied:
The key difference is that in this module, there is no function definition. Bear in mind that druntime is compiled, and distributed as a library.
Hence when the user compile, using `-version=ANSI` won't have any effect on this code.

ErnyTech commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,268 @@
+* Demangles C++ mangled names.
+*
+* If it is not a C++ mangled name or cppdemangle is not supported by your platform, the original mangled C++ name will be returned.
+* The optional destination buffer and return will be contains the same string if the demangle is successful.
```
Ok but the core.demanlge demangle function also uses that name.
Should it be changed?

ErnyTech replied:
Yeah, sorry I used an old description

ErnyTech commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,268 @@
+* The optional destination buffer and return will be contains the same string if the demangle is successful.
+*
+ * Params:
+ *  buf = The string to demangle.
```
Ok but the core.demanlge demangle function also uses that name.
Should it be changed?

Geod24 replied:
Yes

ErnyTech commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,268 @@
+ * Returns:
+ *  The demangled name or the original string if the name is not a mangled C++ name.
+ */
+char[] cppdemangle(const(char)[] buf, char[] dst, bool withPrefix = false, string prefix = "[C++]") @safe
```
So do you think it's better to provide a fixed size buffer like I did with the Windows version?

Geod24 replied:
Yeah, I'd say only accept non-null buffer for the time being

ErnyTech replied:
it's a problem because __cxa_demangle was created to call realloc automatically if the buffer is too small, I have no way of preventing this.
For this reason I passed null to the function which returned me a string pointer of unpredictable size, then I copy the characters in the buffer passed to the function up to the available size in the buffer.

Geod24 replied:
Oh right, read the documentation again and it's not a great behavior. Especially the reliance on `malloc`. Reading again, it *must* be allocated with malloc, which makes the API very unsafe for external user. Can you document this ?

ErnyTech replied:
you are really unsafe, I think the only way is to let it allocate then copy the result (if the buffer has enough space) and free the memory

ErnyTech commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,268 @@
+        }
+    }
+
+    version (Posix)
```
ok

ErnyTech commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,268 @@
+
+    version (Posix)
+    {
+        static ~this() {
```
ok

ErnyTech commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,268 @@
+        return dst[0..(demangle.length + dstStart)];
+    }
+
+    static CPPDemangleStatus inizialize() @trusted
```
ok

ErnyTech commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,268 @@
+                enum names = ["libc++abi.dylib", "libstdc++.dylib"];
+            else version (Posix)
+            {
+                enum names = ["libstdc++.so", "libstdc++.so.6", "libstdc++.so.5",
```
ok

ErnyTech commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,268 @@
+* For Posix platform the demangling is done by function `__cxa_demangle` presents in Itanium C++ ABI.
+* For Windows platform the demangling is done by function `UnDecorateSymbolName` presents in the Debug Help Library
+* If the function is not found or something fails, the original mangled C++ name will be returned.
+*
```
ok

ErnyTech commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,268 @@
+*
+* For Posix platform the demangling is done by function `__cxa_demangle` presents in Itanium C++ ABI.
+* For Windows platform the demangling is done by function `UnDecorateSymbolName` presents in the Debug Help Library
+* If the function is not found or something fails, the original mangled C++ name will be returned.
```
Ok I'll explain it in a better world

ErnyTech commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,168 @@
+/**
+* This module provides routines to demangle C++ symbols
+*
+* For Posix platform the demangling is done by function `__cxa_demangle` presents in Itanium C++ ABI.
```
ok

Commit:
Add cppdemangle in druntime

Having a C++ symbol demangler is very useful, with this it will
be possible to implement the C++ symbol demangle in the stacktrace.

The demangling is done in Posix platform by function __cxa_demangle
presents in Itanium C++ ABI (https://itanium-cxx-abi.github.io/cxx-abi/abi.html#demangler).

Itanium C++ ABI is used practically in all modern C++ compilers on Posix,
however old compilers (like GCC < 3.0) used a different name mangling,
this is not a problem because DMD in fact only supports Itanium C++ ABI on Posix.

For Windows instead the implementation it is provided by the UnDecorateSymbolName
function present in the Debug Help Library.
(https://docs.microsoft.com/en-us/windows/win32/api/dbghelp/nf-dbghelp-undecoratesymbolname)

Signed-off-by: Ernesto Castellotti <mail@ernestocastellotti.it>

ErnyTech commented:
@Geod24 , it's ready for second round ;-)

MoonlightSentinel commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,311 @@
+/**
+* Demangles C++ mangled names.
+*
+* The output buffer and return will be contains the demangled name of C++ symbol if the demangling is successful.
```
```suggestion
* The output buffer and return value will contain the demangled name of C++ symbol if the demangling is successful.
```

MoonlightSentinel commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,311 @@
+*
+* The output buffer and return will be contains the demangled name of C++ symbol if the demangling is successful.
+*
+* This function will fail and the mangled name in input will be returned if:
```
I think the "return input on failure" part should be mentioned in the previous sentence as it describes the functions output.

MoonlightSentinel commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,311 @@
+*
+ * Params:
+ *  mangledName = The string to demangle.
+ *  outputBuffer = The destination buffer, if the size of the destination buffer is <= the size of cppdemangle output the return string would be incomplete.
```
Redundant size requirement, it's already mentioned above

MoonlightSentinel commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,311 @@
+ * Returns:
+ *  The demangled name or the original string if the name is not a mangled C++ name.
+ */
+char[] cppdemangle(const(char)[] mangledName, char[] outputBuffer, string prefix = "") @trusted
```
Don't think this function should accept a prefix, it's an unrelated complication and a user can always do that manually:
```
buffer[0..prefix.length] = prefix;
cppdemangle(mangling, buffer[prefix.length .. $]);

MoonlightSentinel commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,311 @@
+ * Returns:
+ *  The demangled name or the original string if the name is not a mangled C++ name.
+ */
+char[] cppdemangle(const(char)[] mangledName, char[] outputBuffer, string prefix = "") @trusted
```
The symbol should be in camelCase: `cppDemangle`

MoonlightSentinel replied:
`@trusted` should probably be used in smaller scope(s).

MoonlightSentinel commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,311 @@
+    {
+        import core.sys.windows.dbghelp : UnDecorateSymbolNameFunc;
+
+        @nogc nothrow extern(System)
+        {
+            private UnDecorateSymbolNameFunc UnDecorateSymbolName;
+        }
```
`@nogc nothrow extern(System)` doesn't affect the type of `UnDecorateSymbolName` and `extern(System)` is redundant anyway

MoonlightSentinel commented on src/core/internal/cppdemangle.d:
```diff
@@ -0,0 +1,311 @@
+        }
+    }
+
+    version (Windows)
```
```suggestion
    else version (Windows)
```

ErnyTech commented:
Now that my exams are over I can continue with this!
Thanks for the review @Moonlight Sentinel

MoonlightSentinel commented:
Ping @ErnyTech

ljmf00 reviewed (changes requested):
Just a note to tick later: Missing tests.
